<a href="https://colab.research.google.com/github/Jinukki/KJU/blob/master/15108329_ch5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 5장 연습문제
---

# Q1

---

## 다음은 Calculator 클래스이다.

In [0]:
class Calculator:
    def __init__(self):
        self.value=0
        
    def add(self, val):
        self.value+=val

## 위 클래스를 상속하는 UpgradeCalculator를 만들고 값을 뺄 수 있는 minus 메서드를 추가해 보자. 즉 다음과 같이 동작하는 클래스를 만들어야 한다.

cal = UpgradeCalculator()<br>
cal.add(10)<br>
cal.minus(7)<br>

print(cal.value) --- #10에서 7을 뺀 3을 출력

In [0]:
class UpgradeCalculator(Calculator):
    def minus(self,val):
        self.value-=val

In [0]:
cal=UpgradeCalculator()
cal.add(10)
cal.minus(7)

print(cal.value)

3


# Q2

---

## 객체변수 value가 100 이상의 값은 가질 수 없도록 제한하는 MaxLimitCalculator 클래스를 만들어 보자. 즉 다음과 같이 동작해야 한다.

cal = MaxLimitCalculator()<br>
cal.add(50) --- # 50 더하기<br>
cal.add(60) --- # 60 더하기<br>

print(cal.value) --- # 100 출력

## 단 반드시 다음과 같은 Calculator 클래스를 상속해서 만들어야 한다.

In [0]:
class Calculator:
    def __init__(self):
        self.value=0

    def add(self, val):
        self.value+=val

In [0]:
class MaxLimitCalculator(Calculator):
    def add(self, val):
        self.value+=val
        if self.value>100:
            self.value=100

In [0]:
cal = MaxLimitCalculator()
cal.add(50)
cal.add(60)

print(cal.value)

100


# Q3

---

## 다음 결과를 예측해 보자.<br>하나.
all([1, 2, abs(-3)-3])
## 둘.
chr(ord('a')) == 'a'

In [0]:
all([1, 2, abs(-3)-3])

False

### abs(-3)은 -3의 절댓값이므로 3이 되어 all([1, 2, 0])이 되고, 리스트의 요솟값중 0이 있기 때문에 all 내장 함수의 결과는 False가 된다.

In [0]:
chr(ord('a'))=='a'

True

### ord('a') 의 결과는 97이 되어 chr(97)로 치환된다. chr(97)의 결과는 다시 'a'가 되므로 'a' == 'a'가 되어 True를 돌려준다.

# Q4

---

## filter와 lambda를 사용하여 리스트 [1, -2, 3, -5, 8, -3]에서 음수를 모두 제거해 보자.

In [0]:
list(filter(lambda x:x>0, [1, -2, 3, -5, 8, -3]))

[1, 3, 8]

# Q5

---

## 234라는 10진수의 16진수는 다음과 같이 구할 수 있다.

In [0]:
hex(234)

'0xea'

## 이번에는 반대로 16진수 문자열 0xea를 10진수로 변경해 보자.
내장 함수 int를 활용해 보자.

In [0]:
int('0xea', 16)

234

# Q6

---

## map과 lambda를 사용하여 [1, 2, 3, 4] 리스트의 각 요솟값에 3이 곱해진 리스트 [3, 6, 9, 12]를 만들어 보자.

In [0]:
list(map(lambda x:x*3, [1,2,3,4]))

[3, 6, 9, 12]

# Q7

---

## 다음 리스트의 최댓값과 최솟값의 합을 구해 보자.
[-8, 2, 7, 5, -3, 5, 0, 1]

In [0]:
c=[-8, 2, 7, 5, -3, 5, 0, 1]
max(c)+min(c)

-1

# Q8

---

## 17 / 3의 결과는 다음과 같다.

In [0]:
17 / 3

5.666666666666667

## 위와 같은 결괏값 5.666666666666667을 소숫점 4자리까지만 반올림하여 표시해 보자.

In [0]:
round(17/3,4)

5.6667

# Q9

---

## 다음과 같이 실행할 때 입력값을 모두 더하여 출력하는 스크립트(C:\doit\myargv.py)를 작성해 보자.
C:\> cd doit<br>
C:\doit> python myargv.py 1 2 3 4 5 6 7 8 9 10<br>
55

In [7]:
myargv=sys.argv
sys.argv=[myargv, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
import sys

numbers=sys.argv[1:]

result=0

for number in numbers:
    result += int(number)

print(result)

55


# Q10

---

## os 모듈을 사용하여 다음과 같이 동작하도록 코드를 작성해 보자.

1. C:\doit 디렉터리로 이동한다.
2. dir 명령을 실행하고 그 결과를 변수에 담는다.
3. dir 명령의 결과를 출력한다.

In [0]:
import os
os.chdir("D:\\notebooks\doit")

In [0]:
import os
os.getcwd()

'D:\\notebooks\\doit'

In [0]:
result=os.popen("dir")

In [0]:
print(result.read())

 D 드라이브의 볼륨에는 이름이 없습니다.
 볼륨 일련 번호: 36AB-549A

 D:\notebooks\doit 디렉터리

2019-10-09  오전 11:44    <DIR>          .
2019-10-09  오전 11:44    <DIR>          ..
               0개 파일                   0 바이트
               2개 디렉터리  389,987,893,248 바이트 남음



# Q11

---

## glob 모듈을 사용하여 C:\doit 디렉터리의 파일 중 확장자가 .py인 파일만 출력하는 프로그램을 작성해 보자.

## Jupyter Notebook과 연동시킨 폴더로 바꾸어서 진행해보았다. 

In [0]:
import glob
glob.glob("D:/notebooks/*.py")

['D:/notebooks\\buy_apple.py',
 'D:/notebooks\\buy_apple_orange.py',
 'D:/notebooks\\gradient_simplenet.py',
 'D:/notebooks\\layers.py',
 'D:/notebooks\\layer_naive.py',
 'D:/notebooks\\train_neuralnet.py',
 'D:/notebooks\\two_layer_net.py']

# Q12

---

## time 모듈을 사용하여 현재 날짜와 시간을 다음과 같은 형식으로 출력해 보자.

2018/04/03 17:20:32

In [0]:
import time
time.strftime("%Y/%m/%d %H:%M:%S")

'2019/10/08 17:30:52'

### time 모듈의 strftime을 사용

# Q13

---

## random 모듈을 사용하여 로또 번호(1~45 사이의 숫자 6개)를 생성해 보자(단 중복된 숫자가 있으면 안 됨).

In [0]:
import random

result = []
while len(result) < 6:
    num = random.randint(1, 45)
    if num not in result:
        result.append(num)

print(result)

[43, 31, 40, 20, 5, 36]


### random 모듈의 randint를 사용